src/goto-instrument/nondet_static.cpp

```diff
@@ -99,11 +100,11 @@ static void nondet_static(
 
       if(is_nondet_initializable_static(sym, ns))
       {
-        const auto source_location = instruction.source_location();
-        instruction = goto_programt::make_assignment(
-          code_assignt(
-            sym, side_effect_expr_nondett(sym.type(), source_location)),
-          source_location);
+        side_effect_expr_nondett nondet{
+          sym.type(), instruction.source_location()};
+        instruction.assign_rhs_nonconst() = nondet;
+        goto_model.symbol_table.get_writeable_ref(sym.get_identifier()).value =
+          nondet;
```

Comment:
This line here seems to be the cause of the failure in the incremental SMT2 decision procedure. The failure happens during the conversion of the `y` symbol. Before your PR, the symbol value is an expression for the integer 23. After your PR it gains the value `side_effect
  * type: signedbv
      * width: 32
      * #c_type: signed_int
  * #source_location: 
    * file: main.c
    * line: 4
    * function: 
    * working_directory: /home/tspriggs/Development/cbmc/regression/cbmc/no_nondet_static
  * statement: nondet
  * is_nondet_nullable: 1`
This isn't supported by the decision procedure. I think the expressions sent to it either directly or from looking up the values of symbols are expected to be side effect free. The only exception would be in the case where we take the address of a code typed expression. The example in this test does not fall into this exception, because `y` symbol_exprt has a bit vector type.

It could be argued that the decision procedure should not error in this case. However I would say that this change in the value of symbols unrelated to max_malloc_size / object bits appears to be a change unrelated to the intended purpose of this PR.

Reply:
I need to figure out why the value in the symbol table is used directly. What should happen is that such a value is compiled into an assignment in `__CPROVER_initialize`, and `replace_nondet` will resolve this during symex.

Reply:
The problem appears to have been introduced in f286c07c07d: @thomasspriggs, why is it deemed necessary to lookup a symbol's value in `smt2_incremental_decision_proceduret::define_dependent_functions`?!

Reply:
Because the API for `decision_proceduret` is not sufficiently well documented or explicit with regards to how symbol expressions should be treated? I have (incorrectly) assumed that symbol expressions referring to symbols which have not previously been seen, means that these symbols are implicitly defined such that their value is equal to the value in the symbol table. This seemed to work well up to this point.

I can raise a PR which changes this such that symbol expressions are implicitly defined to have a nondeterministic value instead, if that is how a decision procedure should be implemented.